[Blog] Port Kubernetes Meets Event-Driven Ansible to .md (#323)
ansible-community/community-website: posts/2023/03/16/kubernetes-meets-event-driven-ansible.md

---
title: Kubernetes Meets Event-Driven Ansible
author: Andrew Block
date: 2023-03-16 13:00:00
slug: kubernetes-meets-event-driven-ansible
category: event-driven-ansible
tags: event-driven-ansible, kubernetes, rulebooks, eda, ansible-rulebook
type: text
---

> In today’s fast paced world, every second counts and the ability to react to activities in a timely fashion can mean the difference between satisfying the needs of consumers and meeting Service-Level Agreements. Each are goals of [Event-Driven Ansible](https://www.ansible.com/use-cases/event-driven-automation), which seeks to further the reach of Ansible based automation by responding to events that meet certain criteria. These events can originate from a variety of sources, such as from an HTTP endpoint, messages on a queue or topic, or from public cloud resources. Kubernetes has become synonymous with managing infrastructure and applications in cloud native architectures and many organizations are reliant on these systems for running their business critical workloads. Automation and Kubernetes go hand in hand and Ansible already plays a role within this ecosystem. A new capability leveraging the Event-Driven Ansible framework is now available that extends the integration between both Ansible and Kubernetes so that Ansible automation activities can be triggered based on events and actions occurring within a Kubernetes cluster.

Reviewer: ```suggestion
In today's fast paced world, every second counts and the ability to react to activities in a timely fashion can mean the difference between satisfying the needs of consumers and meeting Service-Level Agreements. Each are goals of [Event-Driven Ansible](https://www.ansible.com/use-cases/event-driven-automation), which seeks to further the reach of Ansible based automation by responding to events that meet certain criteria. These events can originate from a variety of sources, such as from an HTTP endpoint, messages on a queue or topic, or from public cloud resources. Kubernetes has become synonymous with managing infrastructure and applications in cloud native architectures and many organizations are reliant on these systems for running their business critical workloads. Automation and Kubernetes go hand in hand and Ansible already plays a role within this ecosystem. A new capability leveraging the Event-Driven Ansible framework is now available that extends the integration between both Ansible and Kubernetes so that Ansible automation activities can be triggered based on events and actions occurring within a Kubernetes cluster.
```

---
title: Kubernetes Meets Event-Driven Ansible
author: Andrew Block
date: 2023-03-16 13:00:00
slug: kubernetes-meets-event-driven-ansible
category: event-driven-ansible
tags: event-driven-ansible, kubernetes, rulebooks, eda, ansible-rulebook
type: text
---

> 
> # __The Inside Playbook__
> 
> ![Kubernetes Meets Event-Driven Ansible](/images/posts/kubernetes-meets-event-driven-ansible/Kubernetes%20+%20EDA.webp)
> 

Reviewer: ```suggestion
```

```
pip install requests
```

Now that all of the prerequisite tooling has been met, attention can be turned to how a Rulebook can be configured to take advantage of the Kubernetes integration. Events in the Event-Driven Ansible architecture are configured within the sources section of a rulebook. One or more sources can be specified within a rulebook enabling a robust set of conditions and actions to be configured.

A basic rulebook that takes advantage of the k8s event source plugin from the Collection is shown below:

```
- name: Listen for newly added ConfigMap resources
  hosts: all
  sources:
    - sabre1041.eda.k8s:
        api_version: v1
        kind: ConfigMap
        namespace: default
  rules:
    - name: Notify
      condition: event.type == "ADDED"
      action:
        debug:  
```

The k8s plugin is modeled in a similar manner to that of the [k8s module](https://docs.ansible.com/ansible/latest/collections/kubernetes/core/k8s_module.html) from the kubernetes.core collection, so anyone with familiarity working with this module will feel at home when working with this k8s source plugin.

> 
> 
> 
> 

Reviewer: ```suggestion
```

---
title: Kubernetes Meets Event-Driven Ansible
author: Andrew Block
date: 2023-03-16 13:00:00
slug: kubernetes-meets-event-driven-ansible
category: event-driven-ansible

> tags: event-driven-ansible

Reviewer 2: ```suggestion
tags: event-driven-ansible, kubernetes, rulebooks, eda, ansible-rulebook
```

- Actions - Triggering the execution of assets including an Ansible Playbook or module 
- Rules - Determination of whether received events match certain conditions 
- Sources - Origination of events from external entities that are consumed and processed within the Ansible eventing framework

There is a wide ecosystem of solutions available to manage Kubernetes from Ansible, which is provided primarily through the [kubernetes.core collection](https://github.com/ansible-collections/kubernetes.core). This Collection contains everything ranging from mechanisms to manage resources within a Kubernetes cluster, support for the Helm package manager to leveraging a Kubernetes cluster as an inventory source. There are capabilities made available now through the integration of Kubernetes and the Event-Driven Ansible framework. Event sources enable the consumption of changes originating from a Kubernetes cluster, which can be used to trigger automation to respond and act based on the received content and the configured rules. Let’s explore how to take advantage of this newly created capability to further the integration between Kubernetes and Ansible.

> The assets related to Event-Driven Ansible and Kubernetes are located within the [sabre1041.eda collection](https://galaxy.ansible.com/sabre1041/eda) within Ansible Galaxy. Ensure that the control node where the Ansible automation will be executed has the necessary tooling installed and configured. This includes Ansible Core, the tooling associated with Event-Driven Ansible, and the Collection containing the Event-Driven Ansible Kubernetes integration. Consult the associated documentation for both [Ansible Core](https://docs.ansible.com/ansible/latest/installation_guide/intro_installation.html) and [Event-Driven Ansible](https://www.ansible.com/blog/getting-started-with-event-driven-ansible) for the target Operating System and installation method.

Reviewer 3: There is a link to the old blog which would break after the migration:
`[Event-Driven Ansible](https://www.ansible.com/blog/getting-started-with-event-driven-ansible)`

It can be replaced later, but mentioning it so we don't forget.

> The assets related to Event-Driven Ansible and Kubernetes are located within the [sabre1041.eda collection](https://galaxy.ansible.com/sabre1041/eda) within Ansible Galaxy. Ensure that the control node where the Ansible automation will be executed has the necessary tooling installed and configured. This includes Ansible Core, the tooling associated with Event-Driven Ansible, and the Collection containing the Event-Driven Ansible Kubernetes integration. Consult the associated documentation for both [Ansible Core](https://docs.ansible.com/ansible/latest/installation_guide/intro_installation.html) and [Event-Driven Ansible](https://www.ansible.com/blog/getting-started-with-event-driven-ansible) for the target Operating System and installation method.

Reviewer: ```suggestion
The assets related to Event-Driven Ansible and Kubernetes are located within the [sabre1041.eda collection](https://galaxy.ansible.com/sabre1041/eda) within Ansible Galaxy. Ensure that the control node where the Ansible automation will be executed has the necessary tooling installed and configured. This includes Ansible Core, the tooling associated with Event-Driven Ansible, and the Collection containing the Event-Driven Ansible Kubernetes integration. Consult the associated documentation for both [Ansible Core](https://docs.ansible.com/ansible/latest/installation_guide/intro_installation.html) and [Event-Driven Ansible](https://access.redhat.com/documentation/en-us/red_hat_ansible_automation_platform/) for the target Operating System and installation method.
```

To demonstrate the use of this rulebook, add the previously provided content to a new rulebook file called k8s-eda-demo.yaml. In addition, ensure that the local machine is authenticated to a Kubernetes cluster or customize the plugin parameters to specify the location of the Kubernetes cluster that should be used. Consult the [plugin documentation](https://github.com/sabre1041/sabre1041.eda/blob/main/docs/sabre1041.eda.k8s_source_plugin.rst) for the available options.

Create a simple [inventory](https://docs.ansible.com/ansible/latest//inventory_guide/intro_inventory.html#intro-inventory) in a file called inventory with the following content:

``` 
localhost
```

Run the rulebook to begin consuming events by executing the following command:

```
ansible-rulebook -i inventory --rulebook k8s-eda-demo.yaml --verbose

```

With the rulebook monitoring for ConfigMap changes in the default namespace, create a new ConfigMap in the default namespace to demonstrate events are being captured appropriately. This task can be accomplished by using the Kubernetes CLI (kubectl) by executing the following command:

 

> 

Reviewer 2: ```suggestion
```
```

> 

Reviewer 2: ```suggestion
~~~
```

> kubectl create configmap -n default eda-example --from-literal=message=”Kubernetes Meets Event-Driven Ansible”

Reviewer 2: ```suggestion
```
kubectl create configmap -n default eda-example --from-literal=message=”Kubernetes Meets Event-Driven Ansible”
```
```

> 

Reviewer 2: ```suggestion
~~~
```

```
ansible-rulebook -i inventory --rulebook k8s-eda-demo.yaml --verbose 
```

Delete and recreate the ConfigMap to trigger the playbook.

```
kubectl delete configmap -n default eda-example

kubectl create configmap -n default eda-example --from-literal=message=”Kubernetes Meets Event-Driven Ansible”
```

Observe that the playbook has been triggered and produces output similar to the following:

```
TASK [ansible.builtin.debug] ***************************************************
ok: [localhost] => {
    "msg": "ConfigMap in namespace 'default' with name 'eda-example ‘Added’ with the message ‘Kubernetes Meets Event-Driven Ansible'"
}
```

While this example only prints out a simple message related to the content of the event received, it provides a demonstration of how to make use of the capabilities enabled by the Kubernetes integration. By adding Kuberentes as an event source and into the Event-Driven Ansible ecosystem, it becomes an essential integration to help support organizations that are leveraging Kubernetes to maintain crucial components of their business and to trigger automation as desired.

__Topics:__

> [Event-Driven Ansible](https://www.ansible.com/blog/topic/event-driven-ansible)

Reviewer 3: We should remove this or change the link to the "event-driven-ansible" tag in the Nikola format:
`https://ansible.com/blog/categories/event-driven-ansible/`

Note: that's assuming we keep the ansible.com/blog path and don't move to a subdomain (not sure it's possible to use Nikola for the main page and a subdomain at the same time)

> 
> __Topics:__
> 
> [Event-Driven Ansible](https://www.ansible.com/blog/topic/event-driven-ansible)

Reviewer: ```suggestion

```

To demonstrate the use of this rulebook, add the previously provided content to a new rulebook file called k8s-eda-demo.yaml. In addition, ensure that the local machine is authenticated to a Kubernetes cluster or customize the plugin parameters to specify the location of the Kubernetes cluster that should be used. Consult the [plugin documentation](https://github.com/sabre1041/sabre1041.eda/blob/main/docs/sabre1041.eda.k8s_source_plugin.rst) for the available options.

Create a simple [inventory](https://docs.ansible.com/ansible/latest//inventory_guide/intro_inventory.html#intro-inventory) in a file called inventory with the following content:

``` 
localhost
```

Run the rulebook to begin consuming events by executing the following command:

```
ansible-rulebook -i inventory --rulebook k8s-eda-demo.yaml --verbose

```

With the rulebook monitoring for ConfigMap changes in the default namespace, create a new ConfigMap in the default namespace to demonstrate events are being captured appropriately. This task can be accomplished by using the Kubernetes CLI (kubectl) by executing the following command:

 

>  ```

Reviewer 2: ```suggestion
```
```

kubectl create configmap -n default eda-example --from-literal=message=”Kubernetes Meets Event-Driven Ansible”

>  ```

Reviewer 2: ```suggestion
```
```
